### PR TITLE
feat(model-apps)!: take up the SDK's injected-storage and async surface

### LINKED ISSUE #582: model-apps: take up the SDK's injected-storage and async surface (parked branch)
Parked work, ready to pick up once #579 lands.

**Branch:** `users/akmaloo/sdk-injected-storage-uptake` (pushed, commit `26e6620e`)
**Scope:** 41 files, +586/−485

## What it does

A **breaking** consumer-side uptake of the cds-maker-sdk changes:

1. The SDK removed `workspacePath`. Every call site builds a store explicitly with
   `createNodeWorkspaceStorage(...)` from the SDK's `/node` entry point, which the vendored bundle
   now re-exports alongside the core. `_vendor-build` bundles a virtual entry re-exporting both
   `./index.js` and `./node/index.js`, so a single `require('./vendor/cds-maker-sdk.cjs')` still
   yields everything the plugin uses.
2. `initWorkspace()` and `createArtifact()` are awaited everywhere — both became promise-returning
   in the SDK's async migration, and an unawaited `createArtifact` does **not** fail loudly: it hands
   back a Promise and the next line reads `.id` as `undefined`.
3. The explicit business-rule validation call is dropped. The SDK now runs the designer's
   completeness validator internally on every save, so an incomplete rule is refused by
   `pushArtifact` rather than by an opt-in check.

## Why it is parked rather than merged

`origin/main` has **not** taken up this change — 31 files there still reference `workspacePath`, and
`createNodeWorkspaceStorage` appears nowhere. So this is genuinely outstanding work, not a
duplicate.

It is held back only to avoid colliding with #579, which is eight commits ahead of `main` on the
same plugin and touches overlapping files.

## Integration notes for whoever picks this up

- ⚠ **Do not cherry-pick it onto #579's branch.** It was built against `origin/main` (2493 tests
  passing there); #579's branch is at 2583. A breaking API migration across 41 files is a real
  integration, not a cherry-pick.
- Rebase onto `main` **after** #579 merges, then re-run the full plugin suite — the test count will
  have moved, and #579 touched `sdk-build.js`, `artifact-intent.js`, `app-spec.js`,
  `download-model-app.js` and `verify-spec.js`, several of which this branch also edits.
- Because it is a breaking change to how every call site constructs the SDK, the real-bundle contract
  tests (`scripts/tests/hardening2-real-bundle.test.js`) are the highest-value gate to re-run first.

### PR DESCRIPTION
**Pull request by akshaymaloo**
Closes #582.

## What this is

The `cds-maker-sdk` removed `workspacePath` and moved the filesystem behind a `/node` entry point. `main` still had **32 `workspacePath` references**, so this is the primary-consumer uptake. Every call site now builds a store explicitly:

```js
const sdk = createMakerSdk({ workspaceStorage: createNodeWorkspaceStorage(dir), instanceUrl, httpClient });
await sdk.initWorkspace();
```

`initWorkspace()` and `createArtifact()` are now awaited everywhere. That matters more than it looks: an unawaited `createArtifact` does **not** fail loudly — it returns a Promise and the next line reads `.id` as `undefined`.

The bundle is built from the **merged** SDK commit (`9719a713a0e9`), which also brings view joins, AI routing description, and refusal of unknown form keys. The earlier draft was blocked on an unmerged commit; that is resolved.

## The defect this PR exists to not ship

Peer review caught a **silent wrong write** that the green suite, the live build, and `--verify` all missed.

The previous bundle consulted `valueWorkflowType` only for `valueType: 'Clear'` and hard-typed every ordinary condition literal as `WorkflowAttributeType.String` (`'14'`). The new bundle uses the field for **all** value types, **verbatim**. `businessRuleDef` was passing the App Spec's human token, and `WorkflowAttributeType` is a numeric-string enum — so the plugin began writing out-of-domain values:

| authored `dataType` | was sent | should be |
| --- | --- | --- |
| *(none)* | `"String"` | `"14"` |
| `Money` | `"Money"` | `"7"` |
| `Picklist` | `"Picklist"` | `"10"` |
| `NotAWorkflowType` | `"NotAWorkflowType"` | rejected |

Note the first row: this hit **every** business rule, not only ones that authored a `dataType`.

**Live A/B settles the consequence.** Same spec, both ways:

| | result |
| --- | --- |
| with the enum mapping | 4/4 rules deployed, `verify PASS 13/13` |
| with the raw token | 4/4 rules deployed, **0 failed**, all activated |

The platform **accepts** the bad value, so this fails **silently**: the rule deploys, activates, and compares with the wrong type. `--verify` cannot see it — it checks existence, duplicate count and `statecode`, never the rule's WFOM.

Two stale claims were corrected rather than left standing: `sdk-build.js` still said `dataType` was *"currently DECORATIVE"*, and `business-rules.test.js` had been rewritten to **pin the pass-through** — asserting `right[0].type === dataType` for a token list that deliberately included `'NotAWorkflowType'`, and reading that as *"dataType IS honoured"*. Pass-through of a non-member value is the opposite of honoured; the test had frozen the defect.

## Also from review: the async guard was blind where the risk moved

`sdk-async-surface.test.js` still listed only the seven originally-async methods, so the one automated guard against unawaited SDK calls did not cover `createArtifact` or `initWorkspace` — the two this uptake made async. Its "still synchronous" counter-examples had been converted to `await sdk.createArtifact(...)` and then asserted not-thenable, which `await` makes **unfalsifiable**: both passed unconditionally while their messages stated the opposite of the truth. Replaced with `assignMissingIds`, one of only three genuinely sync methods left on the bundle (measured), probed without `await`.

## Verification

| Gate | Result |
| --- | --- |
| Plugin suite | 2608/2608 |
| Eval tests | 177/177 |
| Eval runner | 6/6 fixtures, 168 assertions |
| Repository validators | 9/9 |
| `workspacePath` call sites | 0 |

**Live-verified end to end** on a real environment:

- build: 18 created, 0 failed, `verify PASS 19/19`;
- idempotent rebuild: 0 failed, and the deployed **formxml** shows no duplication — tabs `[tab_overview, tab_detail]`, 4 sections, form-columns `60%`/`40%`, `colspan=2`, `rowspan=2`, collapsed tab, 1 sub-grid;
- download round-trip: no Money base-currency twin, no lookup shadow columns, `relationships[]` reconstructed, downloaded spec passes its own lint;
- business rules: 4/4 deployed with mapped enums, `verify PASS 13/13`;
- teardown: zero leftovers, confirmed by query.

Mutation-proven throughout: reverting the enum mapping fails the unit mapping test and the real-bundle wire test; dropping an `await` on a production `createArtifact` now fails the async scan (it did not before); the unknown-form-key test fails against the previous bundle.

## One behaviour change worth knowing

A Picklist condition whose value is the option **label** (`"Closed"`) is now rejected by the platform — `value should be only numeric` — once the condition is correctly typed. The same spec deployed clean when everything was typed String, as a rule that could never match. Loud is better, but it is a change; the schema already documents the numeric option value.

### CODE REVIEW

File: plugins/model-apps/scripts/build-model-app.js [hunk 1]
```diff
@@ -62,15 +62,15 @@ function makeSdk(env, spec, workspaceDir, languageCode) {
...
+    workspaceStorage: createNodeWorkspaceStorage(workspaceDir),
     instanceUrl: env,
     httpClient,
     // Must match the `sdk` instance above: `pushArtifact` refuses a push whose stored artifact
     // language disagrees with the SDK performing it (for language-sensitive registrations), so two
     // instances at different LCIDs would make every push of a fetched artifact fail.
     ...(languageCode ? { languageCode } : {}),
   });
-  provisionSdk.initWorkspace();
+  await provisionSdk.initWorkspace();
```
[the remark this answers is not in the document]
Comment (akshaymaloo) [reply plugins/model-apps/scripts/build-model-app.js]: Confirmed and fixed in `a0919f00`. `cleanup` was defined AFTER the fallible `initWorkspace()`, and the caller's `finally { cleanup() }` only becomes reachable once `makeSdk` **returns** — so an init failure stranded the throwaway workspace for the life of the machine.

One correction to the framing: this is **pre-existing**, not introduced here — `origin/main` has the same shape (`initWorkspace()` at line 73, `const cleanup` at 74). But the uptake genuinely widened the window, which is the part that matters: a rejected async I/O promise is far more reachable than the sync throw it replaced. Worth fixing either way.

Both files now define `cleanup` before the first fallible call and run it in a `catch` before rethrowing — the same pattern `teardown-model-app.js`, `provision-solution.js` and `ai-preflight.js` already used, which is what your comment pointed at.

I checked all five entry points rather than just the two you flagged:

| Entry point | Before | Now |
| --- | --- | --- |
| `build-model-app.js` | leaked | guarded |
| `provision-entities.js` | leaked (2 dirs) | guarded |
| `teardown-model-app.js` | already guarded (try/catch) | unchanged |
| `provision-solution.js` | already guarded (try/finally) | unchanged |
| `ai-preflight.js` | already guarded (try/finally) | unchanged |

`download-model-app.js` is not in the list because its workspace is the caller's output directory, not a throwaway.

No regression test: `makeSdk` is CLI-internal in both files, so exercising the failure would mean exporting it or cache-patching the vendored bundle. I considered a source scan asserting "cleanup defined before init" and rejected it — my own first pass at that heuristic wrongly flagged `teardown-model-app.js`, which defines cleanup after the init but inside a try whose catch cleans up. A check that reports a correct file as broken is worse than none.

File: plugins/model-apps/scripts/provision-entities.js [hunk 2]
```diff
@@ -19,20 +19,20 @@ const { parseArgs, validateFlags, readAliasedFlag, readJsonArg, emitResult, read
...
     instanceUrl: env,
     httpClient,
     solutionUniqueName: input.solution && input.solution.uniqueName,
   });
-  sdk.initWorkspace();
-  const provision = createMakerSdk({ workspacePath: provisionTempDir, instanceUrl: env, httpClient });
-  provision.initWorkspace();
+  await sdk.initWorkspace();
+  const provision = createMakerSdk({ workspaceStorage: createNodeWorkspaceStorage(provisionTempDir), instanceUrl: env, httpClient });
+  await provision.initWorkspace();
```
[the remark this answers is not in the document]
Comment (akshaymaloo) [reply plugins/model-apps/scripts/provision-entities.js]: Confirmed and fixed in `a0919f00`. `cleanup` was defined AFTER the fallible `initWorkspace()`, and the caller's `finally { cleanup() }` only becomes reachable once `makeSdk` **returns** — so an init failure stranded the throwaway workspace for the life of the machine.

One correction to the framing: this is **pre-existing**, not introduced here — `origin/main` has the same shape (`initWorkspace()` at line 73, `const cleanup` at 74). But the uptake genuinely widened the window, which is the part that matters: a rejected async I/O promise is far more reachable than the sync throw it replaced. Worth fixing either way.

Both files now define `cleanup` before the first fallible call and run it in a `catch` before rethrowing — the same pattern `teardown-model-app.js`, `provision-solution.js` and `ai-preflight.js` already used, which is what your comment pointed at.

I checked all five entry points rather than just the two you flagged:

| Entry point | Before | Now |
| --- | --- | --- |
| `build-model-app.js` | leaked | guarded |
| `provision-entities.js` | leaked (2 dirs) | guarded |
| `teardown-model-app.js` | already guarded (try/catch) | unchanged |
| `provision-solution.js` | already guarded (try/finally) | unchanged |
| `ai-preflight.js` | already guarded (try/finally) | unchanged |

`download-model-app.js` is not in the list because its workspace is the caller's output directory, not a throwaway.

No regression test: `makeSdk` is CLI-internal in both files, so exercising the failure would mean exporting it or cache-patching the vendored bundle. I considered a source scan asserting "cleanup defined before init" and rejected it — my own first pass at that heuristic wrongly flagged `teardown-model-app.js`, which defines cleanup after the init but inside a try whose catch cleans up. A check that reports a correct file as broken is worse than none.

File: plugins/model-apps/scripts/tests/sdk-async-surface.test.js [hunk 3]
```diff
@@ -796,7 +796,7 @@ test('the real vendored bundle agrees with ASYNC_SDK_METHODS (no drift in either
     // satisfied. A review pointed out this test proved only one direction; this is the other.
     assert.ok(!(art && typeof art.then === 'function'),
       'createArtifact is still synchronous — the plugin uses its return value directly');
-    const ws = sdk.initWorkspace();
+    const ws = await sdk.initWorkspace();
     assert.ok(!(ws && typeof ws.then === 'function'),
       'initWorkspace is still synchronous — every engine calls it un-awaited before any other work; '
       + 'if it became async, the workspace could be unready when the first artifact call runs');
```
[the remark this answers is not in the document]
Comment (akshaymaloo) [reply plugins/model-apps/scripts/tests/sdk-async-surface.test.js]: Correct, and already fixed in `00db466b` — this review ran against `f72f8fb2`, the first commit on the branch.

An independent review raised the same thing, and it was the sharper half of the finding: **both** "still synchronous" assertions were `await sdk.<method>(...)` followed by `assert.ok(!thenable)`, which `await` makes unfalsifiable. They passed unconditionally while their messages stated the opposite of the measured truth.

Fixed as you describe, plus the part that made them dangerous rather than merely useless: `createArtifact` and `initWorkspace` are now in `ASYNC_SDK_METHODS`, so the await-scan actually covers the two methods this uptake made async and re-awaited at ~10 call sites. It did not before — the tripwire was blind exactly where the risk had moved.

The counter-example is now `assignMissingIds`, probed **without** `await`. I picked it by measuring rather than guessing: calling every method on the bundle and classifying the return leaves only three genuinely synchronous ones (`assertLanguageMatches`, `assertStructurallyValid`, `assignMissingIds`). My first attempt used `setLanguageCode`, which does not exist on the bundle — the `typeof` guard I had written turned the whole check into a silent no-op, so it is now asserted to **exist** as well.

Mutation-proven: dropping an `await` from a production `createArtifact` call now fails the scan, which it did not before this change.

File: plugins/model-apps/scripts/lib/sdk-build.js [hunk 4]
```diff
@@ -2694,45 +2694,29 @@ async function runSdkBuild(spec, opts = {}) {
...
+        const art = await provision.createArtifact('businessRule', def);
         // The condition tree is a nested object, so it goes on through the generic element surface
         // rather than the create payload — mirroring how the SDK's own workflow test authors one.
         await provision.updateElement('businessRule', art.id, '/rootCondition', def.rootCondition);
         // The push CANNOT tell you a rule is wrong. A mis-shaped condition tree is MERGED onto the
         // node, ignored by the serializer, and written as a rule with no clauses and no actions:
         // HTTP 204, activated, and it never fires. That trap is pinned in sdk-uptake-contract.test.js
-        // ("a wrongly-shaped condition produces an EMPTY rule rather than erroring") and, until this
-        // SDK, the spec gate was the only thing standing in front of it.
+        // ("a wrongly-shaped condition produces an EMPTY rule rather than erroring").
```
[the remark this answers is not in the document]
Comment (akshaymaloo) [reply plugins/model-apps/scripts/lib/sdk-build.js]: Right, and it was self-contradictory within the same comment block. Fixed in `a0919f00`.

The retained paragraph claimed *"The push CANNOT tell you a rule is wrong … HTTP 204, activated, and it never fires"* directly above the note explaining that the SDK now refuses exactly that rule at push. It also cited the pinning test by its **old** name (*"produces an EMPTY rule rather than erroring"*); that test was rewritten on this branch to assert the refusal **and** that nothing reached the wire, so the citation pointed at a claim the suite no longer makes.

Measured before rewriting, rather than taking the adjacent paragraph's word for it — an incomplete rule now throws at push with the offending clause named:

```
Validation failed for businessRule/…:
  • /rootCondition (…): Add at least one clause to this condition. [business-rule-EMPTY_CLAUSE_FIELD]
  • /rootCondition (root): A business rule must include at least one action step. [business-rule-NO_ACTION]
```

The comment now describes the trap as **closed at the source**, with the accurate test reference. Your point about it sending a future maintainer toward restoring the deleted validator is the real risk — the surrounding text explicitly argues against re-adding it, so leaving a contradiction beside that argument was the worst of both.

File: plugins/model-apps/scripts/tests/concurrency-stress.test.js [hunk 5]
```diff
@@ -55,24 +55,24 @@ function freshSdk() {
...
 test('STRESS: 40 concurrent artifact mutations all survive in the shared manifest (no lost update)', async () => {
-  const { sdk, dir } = freshSdk();
+  const { sdk, dir } = await freshSdk();
   const N = 40;
 
   // Create first (createArtifact is synchronous), so the concurrent phase is pure async mutation.
   const ids = [];
   for (let i = 0; i < N; i++) {
-    ids.push(sdk.createArtifact('view', { name: `View ${i}`, entityLogicalName: 'account', columns: [] }).id);
+    ids.push((await sdk.createArtifact('view', { name: `View ${i}`, entityLogicalName: 'account', columns: [] })).id);
```
[the remark this answers is not in the document]
Comment (akshaymaloo) [reply plugins/model-apps/scripts/tests/concurrency-stress.test.js]: Correct — fixed in `a0919f00`. The loop did await each creation; only the comment was stale.

Reworded to say what the `await` is actually for: it makes the setup **sequential**, so the phase below is pure concurrent mutation rather than a mix of creates and updates. That is the property the test depends on, and it survives `createArtifact` being async.

**Comment by akshaymaloo** [comment]
## Ready for review — blocker cleared, and peer review caught a silent wrong write

The bundle is now built from the **merged** SDK commit `9719a713a0e9`, so the draft blocker is gone.

### What review found that the green suite did not

| # | Finding | Severity |
| --- | --- | --- |
| 1 | `businessRuleDef` sent the App Spec's word (`Money`, and `String` by default) where Dataverse expects a numeric `WorkflowAttributeType` | **High — silent wrong write** |
| 2 | The async guard did not cover `createArtifact`/`initWorkspace`, and its two counter-assertions were unfalsifiable | Medium |
| 3–7 | Temp-dir leak on failed init (×2), plus three stale comments contradicting the new SDK behaviour | Low–Medium |

Finding 1 is the one that mattered: it hit **every** business rule, not just ones authoring a `dataType`, because the default was the token `'String'` rather than the enum value `'14'`.

**A live A/B settled the consequence** rather than leaving it as "rejects or coerces":

| | result |
| --- | --- |
| enum mapping | 4/4 rules deployed, `verify PASS 13/13` |
| raw token (pre-fix) | 4/4 rules deployed, **0 failed**, all activated |

The platform **accepts** the bad value — so this fails silently, and `--verify` cannot see it (it checks existence, duplicate count and `statecode`, never the rule's WFOM).

Two artefacts had already frozen the defect in place: `sdk-build.js` still said `dataType` was *"currently DECORATIVE"*, and `business-rules.test.js` had been rewritten to **pin the pass-through**, asserting `right[0].type === dataType` for a token list that deliberately included `'NotAWorkflowType'` — and reading that as *"dataType IS honoured"*. Pass-through of a non-member value is the opposite of honoured.

### Live verification

- build: 18 created, 0 failed, `verify PASS 19/19`
- idempotent rebuild: 0 failed; deployed **formxml** shows no duplication — tabs `[tab_overview, tab_detail]`, 4 sections, form-columns `60%`/`40%`, `colspan=2`, `rowspan=2`, collapsed tab, 1 sub-grid
- download round-trip: no Money base-currency twin, no lookup shadows, `relationships[]` reconstructed, downloaded spec passes its own lint
- business rules: 4/4 deployed with mapped enums, `verify PASS 13/13`
- teardown: **zero leftovers**, confirmed by query

### Gates

Plugin 2608/2608 · eval tests 177/177 · eval runner 6/6 fixtures, 168 assertions · validators 9/9 · **CI 10/10** · review threads **5 total, 0 unresolved**, and the latest automated round on `a0919f00` returned **no findings**.

### One behaviour change worth knowing

A Picklist condition whose value is the option **label** is now rejected by the platform (`value should be only numeric`) once the condition is correctly typed. The same spec previously deployed clean — as a rule that could never match. Loud is better, but it is a change.

**Comment by akshaymaloo** [comment]
## Merge order

First half of a two-PR stack — **merge this one first**:

```
main  <--  #590 (this: SDK uptake, breaking + bundle swap)  <--  #592 (layout/verify follow-ups)
```

#592 is based on this branch rather than `main`, so it needs this to land first; GitHub will retarget it to `main` automatically.

They were deliberately kept separate rather than folded together. This PR carries the **breaking** consumer-side change and a 700 KB vendored bundle swap, so keeping it as its own merge commit is what makes a future regression bisectable — and this PR's review is already settled (5 threads resolved, latest automated round clean on `a0919f00`). Combining would have re-opened 44 files of finished review without changing what has been tested, since #592's branch already contains this one in full and every live check there was run on the combined code.

**Comment by akshaymaloo** [comment]
## Live-verified at head `98b6ff48`

The previous live run was at `a0919f00`, before the constructor-guard fix. That fix restructures `makeSdk` — the SDK factory on **every** build path — so re-running live was not optional.

```
build complete — 16 created, 1 skipped, 0 failed (17 steps)
verify PASS (13/13 present)
```

All four business rules deployed with the mapped `WorkflowAttributeType` enums, and teardown removed everything (zero leftovers confirmed by query).

### One failure on the first attempt, and why it is not a regression

The first run failed on the third rule with an HTTP 400 that was a **platform-side SQL exception**, not a validation or payload error:

```
Exception while executing the query:
  insert into DependencyNodeBase(DependencyNodeId, ObjectId, ComponentType, TopSolutionId, ...)
```

That is Dataverse's own solution-dependency bookkeeping, and rules 1 and 2 had just succeeded on the identical code path. I had created and torn down those exact rule names in that exact solution three times today, so residue was the likely cause — but I tested it rather than assuming.

**A/B with the same code and fresh names** (`brwire` → `brfresh`) built all four rules cleanly, `0 failed`, `verify 13/13`. Environment residue, not a code defect. Worth knowing as a live-testing hazard: repeatedly recreating the same business-rule names in the same solution can wedge that internal table in a way teardown cannot reach.

**Review by davidkiz — APPROVED** [review]
